codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/smoketests/SmokeTestsRunnerGenerator.kt

```diff
@@ -0,0 +1,174 @@
+) {
+    internal fun render() {
+        writer.write("import kotlin.system.exitProcess")
+        writer.emptyLine()
+        writer.write("private var exitCode = 0")
+        writer.write("private val regionOverride = System.getenv(\"AWS_SMOKE_TEST_REGION\")")
+        writer.write("private val skipTags = System.getenv(\"AWS_SMOKE_TEST_SKIP_TAGS\")?.let { it.split(\",\").map { it.trim() }.toSet() }")
+        writer.emptyLine()
+        writer.withBlock("public suspend fun main() {", "}") {
+            renderFunctionCalls()
+            write("exitProcess(exitCode)")
+        }
+        writer.emptyLine()
+        renderFunctions()
+    }
+
+    private fun renderFunctionCalls() {
+        operations.forEach { operation ->
+            operation.getTrait<SmokeTestsTrait>()?.testCases?.forEach { testCase ->
+                writer.write("${testCase.id.toCamelCase()}()")
+            }
+        }
+    }
+
+    private fun renderFunctions() {
+        operations.forEach { operation ->
+            operation.getTrait<SmokeTestsTrait>()?.testCases?.forEach { testCase ->
+                renderFunction(operation, testCase)
+                writer.emptyLine()
+            }
+        }
+    }
+
+    private fun renderFunction(operation: OperationShape, testCase: SmokeTestCase) {
+        writer.withBlock("private suspend fun ${testCase.id.toCamelCase()}() {", "}") {
+            write("val tags = setOf<String>(${testCase.tags.joinToString(",") { it.dq()} })")
+            write("if (skipTags != null && tags.any { it in skipTags }) return")
+            emptyLine()
+            withInlineBlock("try {", "} ") {
+                renderClient(testCase)
+                renderOperation(operation, testCase)
+            }
+            withBlock("catch (e: Exception) {", "}") {
+                renderCatchBlock(testCase)
+            }
+        }
+    }
+
+    private fun renderClient(testCase: SmokeTestCase) {
+        writer.withInlineBlock("#L {", "}", service) {
+            if (testCase.vendorParams.isPresent) {
+                testCase.vendorParams.get().members.forEach { vendorParam ->
+                    if (vendorParam.key.value == "region") {
+                        write("region = regionOverride ?: #L", vendorParam.value.render())
+                    } else {
+                        write("#L = #L", vendorParam.key.value.toCamelCase(), vendorParam.value.render())
+                    }
+                }
+            } else {
+                write("region = regionOverride")
+            }
+            val expectingSpecificError = testCase.expectation.failure.getOrNull()?.errorId?.getOrNull() != null
+            write("interceptors.add(#T($expectingSpecificError))", RuntimeTypes.HttpClient.Interceptors.SmokeTestsInterceptor)
+        }
+        checkVendorParamsAreCorrect(testCase)
+    }
+
+    /**
+     * Smithy IDL doesn't check that vendor params are found in vendor params shape so we have to check here.
+     */
```

Comment:
Should this be considered an "Invalid model" and added to the deny list? We shouldn't have to be checking this ourselves